File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/AccountsFragment.java
package com.remotearthsolutions.expensetracker.fragments;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.activities.ApplicationObject;
import com.remotearthsolutions.expensetracker.adapters.AccountsAdapter;
import com.remotearthsolutions.expensetracker.contracts.AccountContract;
import com.remotearthsolutions.expensetracker.contracts.BaseView;
import com.remotearthsolutions.expensetracker.databaseutils.DatabaseClient;
import com.remotearthsolutions.expensetracker.databaseutils.daos.AccountDao;
import com.remotearthsolutions.expensetracker.databaseutils.models.AccountModel;
import com.remotearthsolutions.expensetracker.utils.AlertDialogUtils;
import com.remotearthsolutions.expensetracker.utils.Utils;
import com.remotearthsolutions.expensetracker.viewmodels.AccountViewModel;
import com.remotearthsolutions.expensetracker.viewmodels.viewmodel_factory.AccountViewModelFactory;

import java.util.List;

public class AccountsFragment extends BaseFragment implements AccountContract.View, OptionBottomSheetFragment.Callback {

    private AccountViewModel viewModel;
    private ListView listview;
    private AccountsAdapter adapter;
    private AccountModel selectAccountModel;
    private int limitOfAccount;
    private String currencySymbol;
    private Context context;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
    }

    public AccountsFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_accounts, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        listview = view.findViewById(R.id.accountList);

        currencySymbol = "$";
        if (context != null) {
            currencySymbol = Utils.getCurrency(context);
        }

        AccountDao accountDao = DatabaseClient.getInstance(getContext()).getAppDatabase().accountDao();
        this.viewModel = ViewModelProviders.of(this,
                new AccountViewModelFactory(this, accountDao)).
                get(AccountViewModel.class);
        this.viewModel.loadAccounts();

        viewModel.getNumberOfItem().observe(this,
                (Integer count) -> limitOfAccount = count);

        view.findViewById(R.id.addAccountBtn).setOnClickListener(v -> {

            if (limitOfAccount < 5 ||
                    ((ApplicationObject) ((Activity) context).getApplication()).isPremium()) {
                selectAccountModel = null;
                onClickEditBtn();
            } else {
                showAlert("Attention", "You need to be premium user to add more Account", "Ok", null, null);
            }
        });
    }

    @Override
    public void onAccountFetch(List<AccountModel> accounts) {
        if (isAdded()) {
            adapter = new AccountsAdapter(context, accounts, currencySymbol);
            listview.setAdapter(adapter);
            listview.setOnItemClickListener((parent, view, position, id) -> {
                this.selectAccountModel = accounts.get(position);
                OptionBottomSheetFragment optionBottomSheetFragment = new OptionBottomSheetFragment();
                optionBottomSheetFragment.setCallback(AccountsFragment.this, OptionBottomSheetFragment.OptionsFor.ACCOUNT);
                optionBottomSheetFragment.show(getChildFragmentManager(), OptionBottomSheetFragment.class.getName());
            });
        }
    }

    @Override
    public void onSuccess(String message) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onClickAddAmountBtn() {
        AddAccountAmountDialogFragment addAccountAmountDialogFragment = new AddAccountAmountDialogFragment();
        addAccountAmountDialogFragment.setAccountIncome(selectAccountModel);
        addAccountAmountDialogFragment.setCallback(accountIncome -> {
            viewModel.addOrUpdateAccount(accountIncome);
            addAccountAmountDialogFragment.dismiss();
        });
        addAccountAmountDialogFragment.show(getChildFragmentManager(), AddAccountAmountDialogFragment.class.getName());
    }

    @Override
    public void onClickEditBtn() {
        AddUpdateAccountDialogFragment dialogFragment = new AddUpdateAccountDialogFragment();
        dialogFragment.setAccountModel(selectAccountModel);
        dialogFragment.show(getChildFragmentManager(), AddUpdateAccountDialogFragment.class.getName());

    }

    @Override
    public void onClickDeleteBtn() {
        if (selectAccountModel.getNotremovable() == 1) {
            Toast.makeText(context, "You cannot delete this account", Toast.LENGTH_SHORT).show();
            return;
        }

        AlertDialogUtils.show(context, "Warning",
                "Deleting this account will remove expenses related to this also. Are you sure, You want to Delete?",
                "Yes",
                "Not now",
                new BaseView.Callback() {
                    @Override
                    public void onOkBtnPressed() {
                        viewModel.deleteAccount(selectAccountModel);
                    }

                    @Override
                    public void onCancelBtnPressed() {

                    }
                });
    }

    @Override
    public Context getContext() {
        return context;
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/AccountsFragment.kt
package com.remotearthsolutions.expensetracker.fragments

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.activities.ApplicationObject
import com.remotearthsolutions.expensetracker.activities.main.MainActivity
import com.remotearthsolutions.expensetracker.adapters.AccountsAdapter
import com.remotearthsolutions.expensetracker.contracts.AccountContract
import com.remotearthsolutions.expensetracker.contracts.BaseView
import com.remotearthsolutions.expensetracker.databaseutils.DatabaseClient
import com.remotearthsolutions.expensetracker.databaseutils.models.AccountModel
import com.remotearthsolutions.expensetracker.fragments.OptionBottomSheetFragment.OptionsFor
import com.remotearthsolutions.expensetracker.utils.AlertDialogUtils.show
import com.remotearthsolutions.expensetracker.utils.Constants
import com.remotearthsolutions.expensetracker.utils.SharedPreferenceUtils
import com.remotearthsolutions.expensetracker.utils.Utils.getCurrency
import com.remotearthsolutions.expensetracker.viewmodels.AccountViewModel
import com.remotearthsolutions.expensetracker.viewmodels.viewmodel_factory.BaseViewModelFactory
import kotlinx.android.synthetic.main.fragment_accounts.*
import kotlinx.android.synthetic.main.fragment_accounts.view.*

class AccountsFragment : BaseFragment(),
    AccountContract.View,
    OptionBottomSheetFragment.Callback {
    private var viewModel: AccountViewModel? = null
    private lateinit var mView: View
    private var adapter: AccountsAdapter? = null
    private var selectAccountModel: AccountModel? = null
    private var limitOfAccount = 0
    private var currencySymbol: String? = null
    private var mContext: Context? = null
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        mView = inflater.inflate(R.layout.fragment_accounts, container, false)
        return mView
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)
        currencySymbol = "$"
        if (mContext != null) {
            currencySymbol = getCurrency(mContext!!)
        }

        viewModel =
            ViewModelProviders.of(this, BaseViewModelFactory {
                AccountViewModel(
                    mContext!!, this,
                    DatabaseClient.getInstance(mContext!!).appDatabase.accountDao(),
                    DatabaseClient.getInstance(mContext!!).appDatabase.expenseDao()
                )
            }).get(AccountViewModel::class.java)

        viewModel!!.loadAccounts()
        viewModel!!.numberOfItem.observe(requireActivity(),
            Observer { count: Int -> limitOfAccount = count }
        )
        addAccountBtn.setOnClickListener {
            if (limitOfAccount < 5 ||
                ((mContext as Activity?)!!.application as ApplicationObject).isPremium
            ) {
                selectAccountModel = null
                onClickEditBtn()
            } else {
                showAlert(
                    getString(R.string.attention),
                    getString(R.string.you_need_to_be_premium_user_to_add_more_categories),
                    getString(R.string.ok),
                    null,
                    null
                )
            }
        }

        transferAmountBtn.setOnClickListener {
            val transferBalanceDialogFragment = TransferBalanceDialogFragment()
            transferBalanceDialogFragment.setViewModel(viewModel!!)
            transferBalanceDialogFragment.show(
                childFragmentManager,
                TransferBalanceDialogFragment::class.java.name
            )
        }
    }

    override fun onAccountFetch(accounts: List<AccountModel>?) {
        if (isAdded) {
            adapter = AccountsAdapter(mContext!!, accounts!!, currencySymbol!!)
            mView.accountList.adapter = adapter
            mView.accountList.onItemClickListener =
                AdapterView.OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
                    selectAccountModel = accounts[position]
                    val optionBottomSheetFragment = OptionBottomSheetFragment()
                    optionBottomSheetFragment.setCallback(this@AccountsFragment, OptionsFor.ACCOUNT)
                    optionBottomSheetFragment.show(
                        childFragmentManager,
                        OptionBottomSheetFragment::class.java.name
                    )
                }
        }
    }

    override fun onSuccess(message: String?) {
        Toast.makeText(mContext, message, Toast.LENGTH_SHORT).show()
    }

    override fun onDeleteAccount() {
        (activity as MainActivity).updateSummary()
        if (selectAccountModel?.id!! > 3) {
            SharedPreferenceUtils.getInstance(requireActivity())
                ?.putInt(Constants.KEY_SELECTED_ACCOUNT_ID, 1)
        }
    }

    override fun onClickAddAmountBtn() {
        val addAccountAmountDialogFragment = AddAccountAmountDialogFragment()
        addAccountAmountDialogFragment.setAccountIncome(selectAccountModel)
        addAccountAmountDialogFragment.setCallback(object :
            AddAccountAmountDialogFragment.Callback {
            override fun onAmountAdded(accountIncome: AccountModel?) {
                viewModel!!.addOrUpdateAccount(accountIncome)
                addAccountAmountDialogFragment.dismiss()
            }
        })
        addAccountAmountDialogFragment.show(
            childFragmentManager,
            AddAccountAmountDialogFragment::class.java.name
        )
    }

    override fun onClickEditBtn() {
        val dialogFragment = AddUpdateAccountDialogFragment()
        dialogFragment.setAccountModel(selectAccountModel)
        dialogFragment.show(
            childFragmentManager,
            AddUpdateAccountDialogFragment::class.java.name
        )
    }

    override fun onClickDeleteBtn() {
        if (selectAccountModel!!.notremovable == 1) {
            Toast.makeText(
                activity,
                getString(R.string.you_cannot_delete_this_account),
                Toast.LENGTH_SHORT
            ).show()
            return
        }
        show(activity,
            getString(R.string.warning),
            getString(R.string.deleting_this_account_will_remove_expenses_related_to_this_also_are_you_sure_you_want_to_delete),
            getString(R.string.yes),
            getString(R.string.not_now),
            object : BaseView.Callback {
                override fun onOkBtnPressed() {
                    viewModel!!.deleteAccount(selectAccountModel)
                }

                override fun onCancelBtnPressed() {}
            })
    }
}